File: Src/Runtime/FileSystem/JoinPath.cpp
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/

//
// Created by Rodrigo on 5/29/26.
//

#include "JoinPath.h"

#include <filesystem>

#include "Support/Path/Join.h"
#include "Support/Runtime/AccessString.h"
#include "Support/Runtime/ConvertToString.h"

using namespace Typed;
using namespace Typed::Runtime;
using namespace Typed::Runtime::FileSystem;

ADT::Runtime::Object FileSystem::JoinPath(
    ADT::List::Object &args,
    ADT::Lang::AST *trace
)
{
    auto path = Support::Path::Join(args, trace);

    auto result_str = path.string();
    return {
        ADT::Runtime::ObjectType::String,
        Celery::Str::String{result_str.data(), result_str.size()}
    };
}

File: src/runtime/file_system/join_path.rs
/*
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #                           Typed                     # *
 * #                   A text formatting DSL             # *
 * #                                                     # *
 * #-----------------------------------------------------# *
 * #                                                     # *
 * #         Created by Rodrigo R. & Contributors        # *
 * #         Released under the Apache License 2.0       # *
 * #            Check LICENSE.MD for more info           # *
 * #                                                     # *
 * #-----------------------------------------------------# *
*/
use std::path::PathBuf;
use crate::adt::lang::{ASTValue, RuntimeArguments, AST};
use crate::adt::result::ExecutionResult;
use crate::adt::runtime::{HashableObject, Object};
use crate::support::runtime::object::{get_list, get_string};

pub fn join_path<'a>(
    args: RuntimeArguments<'a>,
    trace: &AST<'a>
)
    -> ExecutionResult<'a>
{
    let list = args.get(0).unwrap();
    let list = get_list(list, trace)?;
    let list = list.borrow();

    let mut base = PathBuf::from("");

    for item in list.iter() {
        let item = get_string(item, trace)?;
        base.push(item);
    }

    let result = base.to_str();
    if result.is_none() {
        return Err(
            crate::adt::error::RuntimeError::could_not_read(trace)
        );
    }

    let base = result.unwrap();
    Ok(
        Object::Hashable(
            HashableObject::String(
                ASTValue::Owned(
                    base.to_owned()
                )
            )
        )
    )
}